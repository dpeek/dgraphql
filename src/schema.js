/* @flow */

import {
  GraphQLID,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLNonNull,
  GraphQLList,
  GraphQLInt,
  parse,
  buildASTSchema,
  Source
} from 'graphql'

import type { GraphQLFieldConfig } from 'graphql'

import { nodeDefinitions } from 'graphql-relay'

import { unwrapNonNull, lowerCamelCase } from './utils'
import { resolveQuery } from './request'

import {
  getDeleteMutation,
  getCreateMutation,
  getUpdateMutation
} from './mutation'

import { getFilterType } from './filter'
import { getOrderType } from './order'
import { getConnectionType } from './connection'

const pluralize: (input: string) => string = require('pluralize')

export type DgraphQLOptions = {
  relay: boolean,
  predicateMap: { [string]: string }
}

function getConnectionField (
  options: DgraphQLOptions,
  type: GraphQLObjectType,
  isRoot: boolean
): GraphQLFieldConfig<{}, {}> {
  const filterType = getFilterType(type)
  const orderType = getOrderType(type)
  return {
    type: options.relay ? getConnectionType(type) : new GraphQLList(type),
    description: `Get filtered, ordered, paginated ${pluralize(type.name)}`,
    args: {
      first: {
        type: GraphQLInt
      },
      after: {
        type: GraphQLID
      },
      ...(filterType && { filter: { type: filterType } }),
      ...(orderType && { order: { type: orderType } })
    },
    resolve: (obj, args, context, info) => {
      if (isRoot) return resolveQuery(options, info)
      if (info.path) return obj[info.path.key]
      return null
    }
  }
}

function getQueryField (
  options: DgraphQLOptions,
  type: GraphQLObjectType
): GraphQLFieldConfig<{}, {}> {
  return {
    type,
    description: `Get a ${type.name} by \`id\``,
    args: {
      id: {
        type: new GraphQLNonNull(GraphQLID)
      }
    },
    resolve: (obj, args, context, info) => resolveQuery(options, info)
  }
}

function transformType (options: DgraphQLOptions, type: GraphQLObjectType) {
  const typeFields = type._typeConfig.fields()
  type._typeConfig.fields = typeFields
  Object.keys(typeFields).forEach(key => {
    const field = typeFields[key]
    const fieldType = unwrapNonNull(field.type)
    if (fieldType instanceof GraphQLList) {
      const nodeType = unwrapNonNull(fieldType.ofType)
      if (nodeType instanceof GraphQLObjectType) {
        const newField = getConnectionField(options, nodeType, false)
        type._typeConfig.fields[key] = newField
      }
    }
  })
  type._fields = null
  return type
}

export function buildSchema (
  graphql: string,
  options: DgraphQLOptions
): GraphQLSchema {
  const predicateMap = {}
  const ast = parse(new Source(graphql))

  ast.definitions.forEach(definition => {
    if (definition.kind !== 'ObjectTypeDefinition') return
    definition.fields.forEach(field => {
      if (!field.directives) return
      field.directives.forEach(directive => {
        if (directive.name.value !== 'dgraph') return
        if (!directive.arguments) return
        directive.arguments.forEach(argument => {
          if (
            argument.value.kind === 'StringValue' &&
            argument.name.value === 'predicate'
          ) {
            predicateMap[definition.name.value + '.' + field.name.value] =
              argument.value.value
          }
        })
      })
    })
  })
  options.predicateMap = predicateMap
  const schema = buildASTSchema(ast)
  const queries = {}
  const mutations = {}
  const typeMap = schema.getTypeMap()

  const node = nodeDefinitions(
    (id, context, info) => resolveQuery(options, info),
    (value, context, info) => info.schema.getType(value.__typename)
  )

  Object.keys(typeMap).forEach(name => {
    if (name.indexOf('_') > -1 || name === 'Query') return
    let type = typeMap[name]
    if (!(type instanceof GraphQLObjectType)) return

    type = transformType(options, type)
    type._interfaces.push(node.nodeInterface)

    let singular = lowerCamelCase(name)
    queries[singular] = getQueryField(options, type)
    queries[pluralize(singular)] = getConnectionField(options, type, true)

    mutations['create' + name] = getCreateMutation(options, type)
    mutations['update' + name] = getUpdateMutation(options, type)
    mutations['delete' + name] = getDeleteMutation(options, type)

    queries['node'] = node.nodeField
  })
  return new GraphQLSchema({
    query: new GraphQLObjectType({
      name: 'Query',
      fields: queries
    }),
    mutation: new GraphQLObjectType({
      name: 'Mutation',
      fields: mutations
    })
  })
}
