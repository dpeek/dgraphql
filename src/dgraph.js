/* @flow */

import fetch from 'isomorphic-fetch'

export type DgraphError = {
  code: 'Error',
  message: string
}

export type DgraphLatency = {
  json: string,
  parsing: string,
  processing: string,
  total: string
}

export type DgraphResult = {
  uids?: { [string]: string },
  server_latency?: DgraphLatency
}

export type DgraphResponse = DgraphResult | DgraphError

function get (server: string, query: string): Promise<DgraphResponse> {
  console.log('-- dgraph query')
  console.log(query)
  return fetch(server + '/query', { method: 'POST', body: query })
    .then(res => res.text())
    .then(res => {
      try {
        return JSON.parse(res)
      } catch (error) {
        return { code: 'Error', message: res }
      }
    })
}

export function connect (server: string) {
  async function query (query: string): Promise<DgraphResult> {
    return get(server, query).then(res => {
      if (res.code === 'Error') throw new Error(res.message)
      return res
    })
  }
  return { query }
}
