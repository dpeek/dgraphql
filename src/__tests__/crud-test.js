import fs from 'fs'
import path from 'path'

import { graphql } from 'graphql'
import { buildSchema } from '../schema'
import { connect } from '../dgraph'

const client = connect('http://localhost:8080')

function getSchema (options) {
  const file = path.resolve(__dirname, 'test.graphql')
  const input = fs.readFileSync(file).toString()
  return buildSchema(input, options)
}

const schema = getSchema({ relay: false })

beforeAll(async () => {
  let dgraphSchemaPath = path.resolve(__dirname, 'test.dgraph')
  await client.query(fs.readFileSync(dgraphSchemaPath).toString())
})

test('creates node', async () => {
  const create = `mutation {
    createPerson(input: {
      name: "Tim"
    }) {
      person {
        id
      }
    }
  }`
  const createResult = await graphql(schema, create)
  const id = createResult.data.createPerson.person.id

  const query = `query {
    person(id: "${id}") {
      name
    }
  }`
  const queryResult = await graphql(schema, query)
  expect(queryResult).toMatchSnapshot()
})

test('returns fields for created node', async () => {
  const create = `mutation {
    createPerson(input: {
      name: "Tim"
    }) {
      person {
        name
      }
    }
  }`
  const createResult = await graphql(schema, create)
  expect(createResult).toMatchSnapshot()
})

test('updates node', async () => {
  const create = `mutation {
    createPerson(input: {
      name: "Tim"
    }) {
      person {
        id
      }
    }
  }`
  const createResult = await graphql(schema, create)
  const id = createResult.data.createPerson.person.id

  const update = `mutation {
    updatePerson(input:{id: "${id}", name: "Jim"}) {
      person {
        name
      }
    }
  }`
  await graphql(schema, update)

  const query = `query {
    person(id: "${id}") {
      name
    }
  }`
  const queryResult = await graphql(schema, query)
  expect(queryResult).toMatchSnapshot()
})

test('returns fields on updated node', async () => {
  const create = `mutation {
    createPerson(input: {
      name: "Tim"
    }) {
      person {
        id
      }
    }
  }`
  const createResult = await graphql(schema, create)
  const id = createResult.data.createPerson.person.id

  const update = `mutation {
    updatePerson(input:{id: "${id}", name: "Jim"}) {
      person {
        name
      }
    }
  }`
  const updateResult = await graphql(schema, update)
  expect(updateResult).toMatchSnapshot()
})

test('deletes node', async () => {
  const create = `mutation {
    createPerson(input: {
      name: "Tim"
    }) {
      person {
        id
      }
    }
  }`
  const createResult = await graphql(schema, create)
  const id = createResult.data.createPerson.person.id

  const deletes = `mutation {
    deletePerson(input: {id: "${id}"}) {
      person {
        id
      }
    }
  }`
  await graphql(schema, deletes)

  const query = `query {
    person(id: "${id}") {
      name
    }
  }`
  const queryResult = await graphql(schema, query)
  expect(queryResult).toMatchSnapshot()
})

test('creates node with nested node', async () => {
  const create = `mutation {
    createPerson(input: {
      name: "Tim",
      partner: {
        name: "Bob"
      }
    }) {
      person {
        id
        partner {
          id
        }
      }
    }
  }`
  const createResult = await graphql(schema, create)
  const personId = createResult.data.createPerson.person.id
  const partnerId = createResult.data.createPerson.person.partner.id

  const query = `query {
    person(id: "${personId}") {
      name
    }
    partner:person(id: "${partnerId}") {
      name
    }
  }`
  const queryResult = await graphql(schema, query)
  expect(queryResult).toMatchSnapshot()
})

test('creates node linked to existing nodes', async () => {
  const create = `mutation {
    mum: createPerson(input: { name: "Linda" }) {
      person {
        id
      }
    }
    dad: createPerson(input: { name: "Matthew" }) {
      person {
        id
      }
    }
  }`
  const createResult = await graphql(schema, create)
  const mumId = createResult.data.mum.person.id
  const dadId = createResult.data.dad.person.id

  const createLinked = `mutation {
    createPerson(
      input: {
        name: "David",
        parents:[
          { id: "${mumId}" },
          { id: "${dadId}" }
        ]
      }
    ) {
      person {
        name
        parents(order: name_asc) {
          name
        }
      }
    }
  }`

  const createLinkedResult = await graphql(schema, createLinked)
  expect(createLinkedResult).toMatchSnapshot()
})
