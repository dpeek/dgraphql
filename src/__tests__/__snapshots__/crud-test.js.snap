// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates node 1`] = `
Object {
  "data": Object {
    "person": Object {
      "name": "Tim",
    },
  },
}
`;

exports[`creates node linked to existing nodes 1`] = `
Object {
  "data": Object {
    "createPerson": Object {
      "person": Object {
        "name": "David",
        "parents": Array [
          Object {
            "name": "Linda",
          },
          Object {
            "name": "Matthew",
          },
        ],
      },
    },
  },
}
`;

exports[`creates node with nested node 1`] = `
Object {
  "data": Object {
    "partner": Object {
      "name": "Bob",
    },
    "person": Object {
      "name": "Tim",
    },
  },
}
`;

exports[`deletes node 1`] = `
Object {
  "data": Object {
    "person": Object {
      "name": null,
    },
  },
}
`;

exports[`returns fields for created node 1`] = `
Object {
  "data": Object {
    "createPerson": Object {
      "person": Object {
        "name": "Tim",
      },
    },
  },
}
`;

exports[`returns fields on updated node 1`] = `
Object {
  "data": Object {
    "updatePerson": Object {
      "person": Object {
        "name": "Jim",
      },
    },
  },
}
`;

exports[`updates node 1`] = `
Object {
  "data": Object {
    "person": Object {
      "name": "Jim",
    },
  },
}
`;
